### DISCUSSION TITLE
[Released] Clone and expose last good deployed commit

### DISCUSSION THREAD

**Discussion by chrisui** [vercel/vercel#7251] [Ideas] [answered]
Right now there is a limitation with the build process (and accompanying ability to ignore specific builds) in that we are a) limited to HEAD^10 commits of history and b) have no idea which the last commit to be built was

These limitations frequently cause false-positive ignored builds, seriously impacting workflows. The only safe workaround right now is to force manual triggering of builds (counter to vercel's ethos) or never ignore any build (wasteful on resources).

## Problems

1. If you push (and trigger a build with) more than one commit at once then `HEAD..HEAD^1` is not correct. Really you want to compare between HEAD and LAST_SUCCESSFUL_COMMIT
2. If you want to always run affected apps between HEAD and main/master you can't unless you're lucky enough to have less than 10 commits on your feature branch.

## Ideas

1. I'd like to propose a new env variable get exposed to the build environment which allows us to know the last good build that we should compare against.
2. I would also like to see a setting which makes checkouts shallow between HEAD and production branch/last good build on another branch (rather than use a magic ^10 number configuration which causes the problems above)
3. Please allow for triggering a build that bypasses the ignore script entirely - "force build" should live alongside "with build cache"

## Context

We have a Nx monorepo and attempting to use vercel to deploy several next apps.

See current recommendation of `--base HEAD^1` (which often doesn't work for reasons mentioned above) [here](https://nx.dev/l/r/guides/deploy-nextjs-to-vercel).

Would like to be able to do the following:

```
npx nx affected:apps --plain --base $DEPLOY_BASE_REF --head HEAD
```

But at the minimum, be able to compare to production branch

```
npx nx affected:apps --plain --base main --head HEAD
# perhaps more portable...
npx nx affected:apps --plain --base $VERCEL_PRODUCTION_REF --head HEAD
```

**Comment by amyegan** [comment, marked as answer]
Thanks for the suggestion! An environment variable is now available to help with this. 😄 

> A new system environment variable: [VERCEL_GIT_PREVIOUS_SHA](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) is now available in the [Ignored Build Step](https://vercel.com/docs/concepts/projects/overview#ignored-build-step), allowing scripts to compare changes against the SHA of the last successful deployment for the current project, and branch.
— [Improvements and Fixes](https://vercel.com/changelog/september-2022-papercuts)
